### ISSUE TITLE
race condition+fail in dist\IO\t\cachepropagate-tcp.t

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12979]
Migrated from [rt.perl.org#118059](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118059) (status was 'resolved')

Searchable as RT118059$

**Comment by p5pRT** [comment]
### From @bulk88

#### Created by @bulk88



There is a race condition in cachepropagate\-tcp\.t between the parent 
proc's accept\(\) and the child proc's connection since the child proc has 
a "sleep\(1\);" delay\. On my Win32 32 bit\, Server 2003 x64\, 8 core\, VC 
2008 Perl\, the \->accept\(\) times out\, does not return an obj\, and then 
fatally errors when \->sockdomain\(\) is called on an undefined scalar\. Example

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
C&#8203;:\\p519\\src\\t>\.\.\\perl\.exe \-I\.\.\\lib harness \.\./dist/IO/t/cachepropagate\-tcp\.t
\.\./dist/IO/t/cachepropagate\-tcp\.t \.\. 1/8 Can't call method "sockdomain" 
on an un
defined value at t/cachepropagate\-tcp\.t line 46\.
\# Looks like you planned 8 tests but ran 5\.
\# Looks like your test exited with 9 just after 5\.
\.\./dist/IO/t/cachepropagate\-tcp\.t \.\. Dubious\, test returned 9 \(wstat 
2304\, 0x900
\)
Failed 3/8 subtests

Test Summary Report
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\.\./dist/IO/t/cachepropagate\-tcp\.t \(Wstat&#8203;: 2304 Tests&#8203;: 5 Failed&#8203;: 0\)
&nbsp;  Non\-zero exit status&#8203;: 9
&nbsp;  Parse errors&#8203;: Bad plan\.  You planned 8 tests but ran 5\.
Files=1\, Tests=5\,  2 wallclock secs \( 0\.03 usr \+  0\.00 sys =  0\.03 CPU\)
Result&#8203;: FAIL

C&#8203;:\\p519\\src\\t>
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

running with \-v\,
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
C&#8203;:\\p519\\src\\t>\.\.\\perl\.exe \-I\.\.\\lib harness \-v 
\.\./dist/IO/t/cachepropagate\-tcp\.t
\.\./dist/IO/t/cachepropagate\-tcp\.t \.\.
1\.\.8
ok 1 \- socket created
ok 2 \- protocol defined
ok 3 \- domain defined
ok 4 \- type defined
ok 5 \- spawned a child
Can't call method "sockdomain" on an undefined value at 
t/cachepropagate\-tcp\.t l
ine 46\.
\# Looks like you planned 8 tests but ran 5\.
\# Looks like your test exited with 9 just after 5\.
Dubious\, test returned 9 \(wstat 2304\, 0x900\)
Failed 3/8 subtests

Test Summary Report
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\.\./dist/IO/t/cachepropagate\-tcp\.t \(Wstat&#8203;: 2304 Tests&#8203;: 5 Failed&#8203;: 0\)
&nbsp;  Non\-zero exit status&#8203;: 9
&nbsp;  Parse errors&#8203;: Bad plan\.  You planned 8 tests but ran 5\.
Files=1\, Tests=5\,  1 wallclock secs \( 0\.02 usr \+  0\.00 sys =  0\.02 CPU\)
Result&#8203;: FAIL

C&#8203;:\\p519\\src\\t>
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

line 46 is
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
&nbsp;    my $new = $listener\->accept\(\);

 >>>>>>>>>>>>>>>>>>    is\($new\->sockdomain\(\)\, $d\, 'domain match'\);
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

accept the function in IO&#8203;::Socket&#8203;::accept the method fails with $\! being 
"Bad file descriptor"/9\. From my research on my machine the timeout is 
undefined\, so IO&#8203;::Select and can\_read parts of IO&#8203;::Socket&#8203;::accept are 
not executed\.

The failing test was added in 
http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/93a5d7bfc07a41ef26fb3e3b298a7d88c3741ed1?f=dist/IO/t/cachepropagate\-tcp\.t 
as part of CPAN RT \#61577 and was written by Daniel Kahn Gillmor\. I dont 
see any explanation for the "sleep\(1\);" in the child fork proc\.

If I put a "sleep\(1\);" before the \->accept\(\)\, it passes for me most \(4 
trys pass\, 5th failed\) of the time\. If I put a 2 sec sleep\, it always 
\(10 trys\, no fail\) passes\, but 2 seconds of sleeping is alot of wall 
time wasted\. If I remove the sleep\(1\) from the child proc\, it always 
fails \(10 fails out of 10 trys\)\. If I put a sleep\(1\) at the accept\, 
remove the sleep\(1\) at the child \(child does not sleep\)\, 10 passes out 
of 10 trys\. This sockets/unix IO/unix events stuff I am not very 
familiar with\, so I dont know how to fix it\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium
    module=IO

Site configuration information for perl 5.19.0:

Configured by Administrator at Sun May 19 19:59:16 2013.

Summary of my perl5 (revision 5 version 19 subversion 0 patch blead 
2013-05-19.21:05:35 bb003204009d113d60d4173c3ed72b10c8169f14 
v5.18.0-25-gbb00320) configuration:
  Snapshot of: bb003204009d113d60d4173c3ed72b10c8169f14
  Platform:
    osname=MSWin32, osvers=5.2, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -GF -W3 -O1 -MD -Zi -DNDEBUG -GS- -GL 
-DWIN32 -D_CONSOLE -DNO_STRICT -D_CRT_SECURE_NO_DEPRECATE 
-D_CRT_NONSTDC_NO_DEPRECATE  -DPERL_TEXTMODE_SCRIPTS 
-DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO',
    optimize='-O1 -MD -Zi -DNDEBUG -GS- -GL',
    cppflags='-DWIN32'
    ccversion='15.00.30729.01', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', 
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf 
-ltcg  -libpath:"c:\p519\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib 
odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib 
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib 
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  
version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl519.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug 
-opt:ref,icf -ltcg  -libpath:"c:\p519\lib\CORE"  -machine:x86'

Locally applied patches:
    


@INC for perl 5.19.0:
    C:/p519/site/lib
    C:/p519/lib
    .


Environment for perl 5.19.0:
    CYGWIN=tty
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=C:\p519\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\Program 
Files (x86)\Microsoft Visual Studio 9.0\VC\bin;C:\Program Files 
(x86)\Microsoft Visual Studio 9.0\VC;C:\Program Files\TortoiseGit\bin
    PERL_BADLANG (unset)
    SHELL (unset)


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun May 19 20&#8203;:24&#8203;:50 2013\, bulk88 wrote&#8203;:
> There is a race condition in cachepropagate\-tcp\.t between the parent
> proc's accept\(\) and the child proc's connection since the child proc
> has
> a "sleep\(1\);" delay\. On my Win32 32 bit\, Server 2003 x64\, 8 core\, VC
> 2008 Perl\, the \->accept\(\) times out\, does not return an obj\, and then
> fatally errors when \->sockdomain\(\) is called on an undefined scalar\.
> Example
> 
>
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
> C&#8203;:\\p519\\src\\t>\.\.\\perl\.exe \-I\.\.\\lib harness
> \.\./dist/IO/t/cachepropagate\-tcp\.t
> \.\./dist/IO/t/cachepropagate\-tcp\.t \.\. 1/8 Can't call method
> "sockdomain"
> on an un
> defined value at t/cachepropagate\-tcp\.t line 46\.
> \# Looks like you planned 8 tests but ran 5\.
> \# Looks like your test exited with 9 just after 5\.
> \.\./dist/IO/t/cachepropagate\-tcp\.t \.\. Dubious\, test returned 9 \(wstat
> 2304\, 0x900
> \)
> Failed 3/8 subtests

I see this failure maybe 1 in 20 runs \(Windows 7 x64\, Core i7\)\.

I suspect the problem is that winsock is invalidating the accept socket 
once the child socket goes out of scope\, causing a race in the accept\(\)
implementation\.

I changes the code so the child reads a line from the socket\, and 
the parent closes the socket\.  With that I had 2700 successful runs
before I stopped testing\.

I'll try a smoke\-me to see if it breaks anywhere else\.

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @kmx

I get the same failure with perl\-5\.18\.0 \+ gcc\-4\.7\.3 \+ 32bit MS Windows
on my Windows 7 box\.

It fails always

\-\-
kmx

**Comment by p5pRT** [comment]
### From @bulk88

On Mon May 20 02&#8203;:10&#8203;:13 2013\, tonyc wrote&#8203;:
> 
> I see this failure maybe 1 in 20 runs \(Windows 7 x64\, Core i7\)\.
> 
> I suspect the problem is that winsock is invalidating the accept socket 
> once the child socket goes out of scope\, causing a race in the accept\(\)
> implementation\.
> 
> I changes the code so the child reads a line from the socket\, and 
> the parent closes the socket\.  With that I had 2700 successful runs
> before I stopped testing\.
> 
> I'll try a smoke\-me to see if it breaks anywhere else\.
> 
> Tony

Any updates TonyC?

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu\, Jun 27\, 2013 at 04&#8203;:34&#8203;:17PM \-0700\, bulk88 via RT wrote&#8203;:
> On Mon May 20 02&#8203;:10&#8203;:13 2013\, tonyc wrote&#8203;:
> > 
> > I see this failure maybe 1 in 20 runs \(Windows 7 x64\, Core i7\)\.
> > 
> > I suspect the problem is that winsock is invalidating the accept socket 
> > once the child socket goes out of scope\, causing a race in the accept\(\)
> > implementation\.
> > 
> > I changes the code so the child reads a line from the socket\, and 
> > the parent closes the socket\.  With that I had 2700 successful runs
> > before I stopped testing\.
> > 
> > I'll try a smoke\-me to see if it breaks anywhere else\.
> > 
> > Tony
> 
> Any updates TonyC?

It smoked ok\, but I don't trust that it fixes the underlying problem
\(which may not be fixable\, leaving us with the work\-around\.\)

My theory above\, as written\, is hopefully nonsense \- Windows returning
a socket and suddenly making it not\-a\-socket\, rather than return an
end\-of\-file or EPIPE on the next operation would be even more broken
than I expect from Microsoft\.

For a Real™ fork\(\) any open file handles or sockets are cloned in the
child \- the child can exit or explicitly cose their socket handle\, but
it won't have an effect on the socket handle the parent has\.

Under Win32 we emulate that\, which I suspect is the real cause of the
problem here \- when a thread is created fp\_dup\(\) calls
PerlIO\_fdupopen\(\) which does reasonable things on Unix\, but on Win32
that leaves all the work to win32\_fdupopen\(\)\.

win32\_fdupopen\(\) calls win32\_dup\(\) a trivial wrapper around dup\(\) \-
and I don't see how that works for a socket fd unless the CRT dup\(\) is
tolerant of errors from DuplicateHandle\(\)\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu Jun 27 23&#8203;:50&#8203;:22 2013\, tonyc wrote&#8203;:
> It smoked ok\, but I don't trust that it fixes the underlying problem
> \(which may not be fixable\, leaving us with the work\-around\.\)
> 
> My theory above\, as written\, is hopefully nonsense \- Windows returning
> a socket and suddenly making it not\-a\-socket\, rather than return an
> end\-of\-file or EPIPE on the next operation would be even more broken
> than I expect from Microsoft\.
> 
> For a Real™ fork\(\) any open file handles or sockets are cloned in the
> child \- the child can exit or explicitly cose their socket handle\, but
> it won't have an effect on the socket handle the parent has\.
> 
> Under Win32 we emulate that\, which I suspect is the real cause of the
> problem here \- when a thread is created fp\_dup\(\) calls
> PerlIO\_fdupopen\(\) which does reasonable things on Unix\, but on Win32
> that leaves all the work to win32\_fdupopen\(\)\.
> 
> win32\_fdupopen\(\) calls win32\_dup\(\) a trivial wrapper around dup\(\) \-
> and I don't see how that works for a socket fd unless the CRT dup\(\) is
> tolerant of errors from DuplicateHandle\(\)\.

This theory turned out to be nonsense\, I'm exploring the behaviour some
more\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue Aug 27 16&#8203;:41&#8203;:11 2013\, tonyc wrote&#8203;:
> On Thu Jun 27 23&#8203;:50&#8203;:22 2013\, tonyc wrote&#8203;:
> > It smoked ok\, but I don't trust that it fixes the underlying problem
> > \(which may not be fixable\, leaving us with the work\-around\.\)
> > 
> > My theory above\, as written\, is hopefully nonsense \- Windows returning
> > a socket and suddenly making it not\-a\-socket\, rather than return an
> > end\-of\-file or EPIPE on the next operation would be even more broken
> > than I expect from Microsoft\.
> > 
> > For a Real™ fork\(\) any open file handles or sockets are cloned in the
> > child \- the child can exit or explicitly cose their socket handle\, but
> > it won't have an effect on the socket handle the parent has\.
> > 
> > Under Win32 we emulate that\, which I suspect is the real cause of the
> > problem here \- when a thread is created fp\_dup\(\) calls
> > PerlIO\_fdupopen\(\) which does reasonable things on Unix\, but on Win32
> > that leaves all the work to win32\_fdupopen\(\)\.
> > 
> > win32\_fdupopen\(\) calls win32\_dup\(\) a trivial wrapper around dup\(\) \-
> > and I don't see how that works for a socket fd unless the CRT dup\(\) is
> > tolerant of errors from DuplicateHandle\(\)\.
> 
> This theory turned out to be nonsense\, I'm exploring the behaviour some
> more\.

Amongst many other things I tried\, I changed win32\_accept to&#8203;:

win32\_accept\(SOCKET s\, struct sockaddr \*addr\, int \*addrlen\)
\{
&nbsp;  SOCKET r\, s2;
&nbsp;  SOCKET x;

&nbsp;    SOCKET\_TEST\(\(r = accept\(TO\_SOCKET\(s\)\, addr\, addrlen\)\)\, INVALID\_SOCKET\);
&nbsp;    if \(r == INVALID\_SOCKET\) \{
&nbsp;      dTHX;
&nbsp;      PerlIO\_printf\(PerlIO\_stderr\(\)\, "accept\(%d \(%p\)\) => %d failed %d\\n"\,
\(int\)s\, \_get\_osfhandle\(s\)\, \(int\)r\, errno\);
&nbsp;    \}
&nbsp;        s2 = OPEN\_SOCKET\(r\);
&nbsp;    x = \_get\_osfhandle\(s2\);
&nbsp;    if \(x \!= r\) \{
&nbsp;      dTHX;
&nbsp;      PerlIO\_printf\(PerlIO\_stderr\(\)\, "accept\(%d \(%d\)\) => %d but osfhandle
returned bad %d\\n"\, \(int\)s\, \_get\_osfhandle\(s\)\, \(int\)r\, \(int\)x\);
&nbsp;    \}
&nbsp;    return s2;
\}

The output from a failed run looked like&#8203;:

1\.\.8
ok 1 \- socket created
ok 2 \- protocol defined
ok 3 \- domain defined
ok 4 \- type defined
ok 5 \- spawned a child
accept\(3 \(112\)\) => 236 but osfhandle returned bad \-1
Can't use an undefined value as a symbol reference at
\.\./dist/IO/t/cachepropagate\-tcp\.t line 49\.
\# Looks like you planned 8 tests but ran 5\.
\# Looks like your test exited with 9 just after 5\.

So it seems some race is closing the accept\(\)ed socket before we get to
use it\.

Pretty much any output I do in the child or parent before the accept\(\)
prevents the problem from occuring for me\, which makes it difficult to test\.

Tony

**Comment by p5pRT** [comment]
### From @bulk88

On Wed Aug 28 18&#8203;:09&#8203;:32 2013\, tonyc wrote&#8203;:
> 
> The output from a failed run looked like&#8203;:
> 
> 1\.\.8
> ok 1 \- socket created
> ok 2 \- protocol defined
> ok 3 \- domain defined
> ok 4 \- type defined
> ok 5 \- spawned a child
> accept\(3 \(112\)\) => 236 but osfhandle returned bad \-1
> Can't use an undefined value as a symbol reference at
> \.\./dist/IO/t/cachepropagate\-tcp\.t line 49\.
> \# Looks like you planned 8 tests but ran 5\.
> \# Looks like your test exited with 9 just after 5\.
> 
> So it seems some race is closing the accept\(\)ed socket before we get to
> use it\.
> 
> Pretty much any output I do in the child or parent before the accept\(\)
> prevents the problem from occuring for me\, which makes it difficult to
test\.
> 
> Tony

Using freeze/thaw features of the parent and child OS threads\, the bug
can be prevented\.  2 theories I have

1\. somewhere a CloseHandle was done on a socket handle\, which isn't
allowed on paper\, because supposedly there are winsock user\-mode
resources which aren't cleaned up when the kernel handle \(a socket on NT
is a kernel handle from the AFD driver\, IDK if its true that the AFD
driver won't cleanup/do a callback to the usermode winsock side when the
kernel handle is closed\) is closed\.

2\. a double free \(using the correct closesocket\(\) command\) was done\,
when the child psuedo proc exited

I include 2 screen shots of why I think the above\. The break happened at
a invalid handle exception from zwclose/ntclose\. Notice it came from the
usermode winsock dll\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @bulk88

On Fri Sep 20 01&#8203;:14&#8203;:51 2013\, bulk88 wrote&#8203;:

> I include 2 screen shots of why I think the above\. The break happened at
> a invalid handle exception from zwclose/ntclose\. Notice it came from the
> usermode winsock dll\.

forgot the attachments

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @bulk88

[118059-9-20-2.GIF](https://rt-archive.perl.org/perl5/Ticket/Attachment/1256791/658365/118059-9-20-2.GIF)

**Comment by p5pRT** [comment]
### From @bulk88

[118059-9-20-1.GIF](https://rt-archive.perl.org/perl5/Ticket/Attachment/1256791/658371/118059-9-20-1.GIF)

**Comment by p5pRT** [comment]
### From @bulk88

Adding a trimmed down test version of cacheproagate\-tcp\.t to isolate
problem\. Another bug investigating is\, in badrun\, the $^E is 6\, which is
ERROR\_INVALID\_HANDLE\, the accept returned 10038/0x2736/WSAENOTSOCK to
perl\. There is a line that must be commented out to be a badrun\,
uncomment \(and the print in the child thread adds delay\) it will be a
good run\. 

Shouldn't $^E be 10038 not 6 at this point? Is this another "bug" or
not? Ill have more to say tomorrow from rewriting a IRC p5p convo today\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @bulk88

1\.\.6
ok 1 \- socket created
ok 2 \- protocol defined
ok 3 \- domain defined
ok 4 \- type defined
listener socket is OS Hnd 368
Press any key to continue \. \. \. 
ok 5 \- spawned a child
client socket os hnd >2d4\< fileno >4\<
rawerr pass cerr >25\< winerr >0\<
$VAR1 = \{
&nbsp;          'SO\_ERROR' => 4103\,
&nbsp;          'SOL\_SOCKET' => 65535\,
&nbsp;          '$listener' => bless\( \\\*Symbol&#8203;::GEN0\, 'IO&#8203;::Socket&#8203;::INET' \)\,
&nbsp;          'SockOptRes' => 0
&nbsp;        \};
IsPerlIOValid\($listener\) 1
post accept cerr >25\< winerr >6\< child OS Hnd >2d4\< parent accept OS Hnd >2c8\<
child fds >4\< parent fds >5\<
ok 6 \- domain match
Press any key to continue \. \. \.

**Comment by p5pRT** [comment]
### From @bulk88

[tcphandlebugs.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1258455/659503/tcphandlebugs.pl)

**Comment by p5pRT** [comment]
### From @bulk88

1\.\.6
ok 1 \- socket created
ok 2 \- protocol defined
ok 3 \- domain defined
ok 4 \- type defined
listener socket is OS Hnd 368
Press any key to continue \. \. \. 
ok 5 \- spawned a child
rawerr pass cerr >9\< winerr >0\<
$VAR1 = \{
&nbsp;          '$listener' => bless\( \\\*Symbol&#8203;::GEN0\, 'IO&#8203;::Socket&#8203;::INET' \)\,
&nbsp;          'SOL\_SOCKET' => 65535\,
&nbsp;          'SO\_ERROR' => 4103\,
&nbsp;          'SockOptRes' => 0
&nbsp;        \};
IsPerlIOValid\($listener\) 1
post accept cerr >9\< winerr >6\< child OS Hnd >2d4\< parent accept OS Hnd >1\.\#QNAN\<
child fds >4\< parent fds >$new is undef\<
sleep block Excpt&#8203;: "Can't call method "sockdomain" on an undefined value at tcphandlebugs\.pl line 75\." 2d4child fds 4
Press any key to continue \. \. \.

**Comment by p5pRT** [comment]
### From @bulk88

On Wed Sep 25 23&#8203;:42&#8203;:55 2013\, bulk88 wrote&#8203;:
> Adding a trimmed down test version of cacheproagate\-tcp\.t to isolate
> problem\. Another bug investigating is\, in badrun\, the $^E is 6\, which is
> ERROR\_INVALID\_HANDLE\, the accept returned 10038/0x2736/WSAENOTSOCK to
> perl\. There is a line that must be commented out to be a badrun\,
> uncomment \(and the print in the child thread adds delay\) it will be a
> good run\. 
> 

Made an ithreads version which freezes the server thread that does the
accept in the accept\, for the whole life time of the child thread I
think\, \(didnt verify with c debugger\)\. It does NOT fail on it\. And the
accept returns a working handle\.

Is this a handle leak in Perl PLUS a race in winsock?

The script can be messed around with\.
\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @bulk88

[tcphandlebugst.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1258521/659555/tcphandlebugst.pl)

**Comment by p5pRT** [comment]
### From @bulk88

Adding some things I said in IRC for archival reasons\.

Would this make any sense\, in Winsock\, if a process does a loopback
connection\, the Winsock creates a pipe OS handle instead of a packetized
TCP OS handle\, then registers the pipe OS handle with a Winsock user
mode struct so all socket funcs will work as normal on the pipe handle\,
then the socket\(\) returns the same handle that accept\(\) returns\, and
instead of going through the OS packet switcher\, they talk over a pipe
which is more efficient? About the bug\, what happens is\, the accept\(\)
calls into kernel mode and the server is descheduled from the CPU\, then
the child psuedo fork runs\, does its connect\(\)\, then closesocket\(\)s
during perl ithread glboal destruction\, then the accept\(\) unblocks from
kernel mode\, but the handle is that unblocked the accept is dead already?

Invalid handle exception with C debugger\.
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
&nbsp;     ntdll\.dll\!\_ZwClose@&#8203;4\(\)  \+ 0x12 bytes    
&nbsp;     mswsock\.dll\!\_SockCloseSocket@&#8203;4\(\)  \+ 0x240 bytes    
&nbsp;     mswsock\.dll\!\_WSPAccept@&#8203;24\(\)  \+ 0xaef bytes    
&nbsp;     ws2\_32\.dll\!\_WSAAccept@&#8203;20\(\)  \+ 0x85 bytes    
&nbsp;     ws2\_32\.dll\!\_accept@&#8203;12\(\)  \+ 0x17 bytes    
>    perl519\.dll\!win32\_accept\(unsigned int s=3\, sockaddr \* addr=0x0012f700\,
int \* addrlen=0x0012fb00\)  Line 370 \+ 0x12 bytes    C
&nbsp;     perl519\.dll\!PerlSockAccept\(IPerlSock \* piPerl=0x01b67218\, unsigned int
s=3\, sockaddr \* addr=0x0012f700\, int \* addrlen=0x0012fb00\)  Line 1227 \+
0x11 bytes    C\+\+
&nbsp;     perl519\.dll\!Perl\_pp\_accept\(interpreter \* my\_perl=0x01b64e0c\)  Line
2557 \+ 0x37 bytes    C
&nbsp;     perl519\.dll\!Perl\_runops\_standard\(interpreter \* my\_perl=0x01b64e0c\) 
Line 42 \+ 0xc bytes    C
&nbsp;     perl519\.dll\!S\_run\_body\(interpreter \* my\_perl=0x01b64e0c\, long
oldscope=1\)  Line 2500 \+ 0xf bytes    C
&nbsp;     perl519\.dll\!perl\_run\(interpreter \* my\_perl=0x01b64e0c\)  Line 2419    C
&nbsp;     perl519\.dll\!RunPerl\(int argc=2\, char \* \* argv=0x01b64d68\, char \* \*
env=0x01b63430\)  Line 270 \+ 0x9 bytes    C\+\+
&nbsp;     perl\.exe\!\_\_tmainCRTStartup\(\)  Line 582 \+ 0x17 bytes    C
&nbsp;     kernel32\.dll\!\_BaseProcessStart@&#8203;4\(\)  \+ 0x28 bytes    
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

The invalid handle passed to NtClose in the parent thread is the same OS
kernel handle that appeared on the CRT level from $connector\.
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
&nbsp;    my $connector = IO&#8203;::Socket&#8203;::INET\->new\(PeerAddr => '127\.0\.0\.1'\,
&nbsp;                          PeerPort => $port\,
&nbsp;                          Proto => 'tcp'\);
&nbsp;    Win32&#8203;::API&#8203;::WriteMemory\($ptr\, pack\('I'\,
Win32API&#8203;::File&#8203;::GetOsFHandle\(\*\{$connector\}\{IO\}\)\)\, 4\);
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
So Winsock somehow lets accept\(\) and its callees in parent thread see
the OS handle of the child thread and try to close it from the parent
thread\.

The pointer memory write stuff was the fastest way I though of for the
child thread to report its data without doing stdio to console which
then stops the race/bug\.

The accept actually fails with 10038 WSAENOTSOCK but PerlIO corrupts it
to ERROR\_INVALID\_HANDLE/6\. Disassembly of WSPAccept and more debugging
shows the 10038 is coming from a NtDeviceIoControlFile call in WSPAccept
with IOCTL 0x12010 which google and reactos say is a constant called
AFD\_ACCEPT\. NtDeviceIoControlFile AFD\_ACCEPT returns 0xc0000008 
STATUS\_INVALID\_HANDLE \. The asm shows SockCloseSocket branch is only
taken on error codes \!= 0\, so its resource cleanup code on failure\.
NtStatusToSocketError later in WSPAccept converts STATUS\_INVALID\_HANDLE
to WSAENOTSOCK which becomes the public facing error code of the failed
accept\(\)\.

Because of the is a loopback in the same process implemented using pipes
by winsock theory I present above\. TonyC asked are the handles the same
from the accept\(\) in the parent thread and connect\(\) in the child
thread? On a successful accept\(\)\, the OS handle returned is DIFFERENT
from the OS handle in $connector in the child\. Not in IRC comment&#8203;: I
didnt research if the 2 handles are the same kernel object/struct in
object manager and Winsock is doing a DuplicateHandle internally\.

This is an answer for another TonyC question\. The fds in a successful
accept\(\) are different also\. There is a class of bugs/limitations in
Win32 psuedo\-fork where the fileno/fds aren't virtualized between the
psuedo\-procs\. Child socket is fd 4 and parent socket is fd 5 in perl
lang on a successful accept\(\) by my tests\. See earlier posts in this ticket\.

**Comment by p5pRT** [comment]
### From @steve-m-hay

Fixed by commit b47a847f62 from \#120091\, so closing this ticket too :\-\)

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'
